modules/apps/transfer/keeper/relay.go

```diff
@@ -214,72 +226,115 @@ func (k Keeper) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, data t
-			return errorsmod.Wrapf(err, "failed to send coins to receiver %s", receiver.String())
+			denomPath := token.Denom.Path()
+			labels = append(labels, telemetry.NewLabel(coretypes.LabelSource, "false"))
+			defer internaltelemetry.ReportOnRecvPacketTelemetry(transferAmount, denomPath, labels)
+
+			receivedCoins = append(receivedCoins, voucher)
 		}
+	}
 
-		denomPath := token.Denom.Path()
-		labels = append(labels, telemetry.NewLabel(coretypes.LabelSource, "false"))
-		defer internaltelemetry.ReportOnRecvPacketTelemetry(transferAmount, denomPath, labels)
+	if data.HasForwarding() {
+		// we are now sending from the forward escrow address to the final receiver address.
+		if err := k.forwardPacket(ctx, data, packet, receivedCoins); err != nil {
+			return err
+		}
 	}
 
+	// The ibc_module.go module will return the proper ack.
 	return nil
 }
 
-// OnAcknowledgementPacket responds to the success or failure of a packet
-// acknowledgement written on the receiving chain. If the acknowledgement
-// was a success then nothing occurs. If the acknowledgement failed, then
-// the sender is refunded their tokens using the refundPacketTokens function.
+// OnAcknowledgementPacket either reverts the state changes executed in receive
+// and send packet if the chain acted as a middle hop on a multihop transfer; or
+// responds to the success or failure of a packet acknowledgement written on the
+// final receiving chain, if it acted as the original sender chain. If the
+// acknowledgement was a success then nothing occurs. If the acknowledgement failed,
+// then the sender is refunded their tokens using the refundPacketToken function.
 func (k Keeper) OnAcknowledgementPacket(ctx sdk.Context, packet channeltypes.Packet, data types.FungibleTokenPacketDataV2, ack channeltypes.Acknowledgement) error {
+	prevPacket, isForwarded := k.getForwardedPacket(ctx, packet.SourcePort, packet.SourceChannel, packet.Sequence)
+
 	switch ack.Response.(type) {
 	case *channeltypes.Acknowledgement_Result:
+		if isForwarded {
+			return k.ackForwardPacketSuccess(ctx, prevPacket, packet)
+		}
+
 		// the acknowledgement succeeded on the receiving chain so nothing
 		// needs to be executed and no error needs to be returned
 		return nil
 	case *channeltypes.Acknowledgement_Error:
-		return k.refundPacketTokens(ctx, packet, data)
+		// We refund the tokens from the escrow address to the sender
+		if err := k.refundPacketTokens(ctx, packet, data); err != nil {
+			return err
+		}
+		if isForwarded {
+			return k.ackForwardPacketError(ctx, prevPacket, packet, data)
+		}
+
+		return nil
 	default:
 		return errorsmod.Wrapf(ibcerrors.ErrInvalidType, "expected one of [%T, %T], got %T", channeltypes.Acknowledgement_Result{}, channeltypes.Acknowledgement_Error{}, ack.Response)
 	}
 }
 
-// OnTimeoutPacket refunds the sender since the original packet sent was
-// never received and has been timed out.
+// OnTimeoutPacket either reverts the state changes executed in receive and send
```

Comment:
nit: I'd vote for godocs to specify ordinary transfer logic first, and then state what happens if its a multihop transfer

can clean these up post merge

Reply:
created https://github.com/cosmos/ibc-go/issues/6724